Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -30,49 +30,87 @@
 #include "Document.h"
 #include "DocumentInlines.h"
 #include "EventNames.h"
+#include "MainThreadPermissionObserver.h"
 #include "PermissionController.h"
+#include "PermissionState.h"
+#include "Permissions.h"
+#include "ScriptExecutionContext.h"
 #include "SecurityOrigin.h"
+#include "WorkerGlobalScope.h"
+#include "WorkerLoaderProxy.h"
+#include "WorkerThread.h"
+#include <memory>
+#include <wtf/HashMap.h>
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/MainThread.h>
+#include <wtf/NeverDestroyed.h>
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(PermissionStatus);
 
-Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+static HashMap<MainThreadPermissionObserverIdentifier, std::unique_ptr<MainThreadPermissionObserver>>& allMainThreadPermissionObservers()
 {
-    auto status = adoptRef(*new PermissionStatus(context, state, descriptor));
+    static MainThreadNeverDestroyed<HashMap<MainThreadPermissionObserverIdentifier, std::unique_ptr<MainThreadPermissionObserver>>> map;
+    return map;
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
+{
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
     status->suspendIfNeeded();
     return status;
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
     : ActiveDOMObject(&context)
     , m_state(state)
     , m_descriptor(descriptor)
+    , m_source(source)
+    , m_page(page)
 {
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
```

Comment:
Would be nice to add such an assertion:
ASSERT(isMainThread() || !page);

The code wouldn't be safe if this assertion didn't hold.

Reply:
But if the source is a Dedicated worker, then we would not be on the main thread and we would have a page. So this assertion would be hit, even though there's nothing wrong no?

Reply:
You were constructing a `WeakPtr<Page>` from a `Page*`. This was only safe on the main thread since the Page gets destroyed on the main thread.